addons/gamification/models/gamification_karma_tracking.py

```diff
@@ -1,69 +1,128 @@
-        AND tracking_date::timestamp BETWEEN %(from_date)s AND %(to_date)s
-        AND consolidated IS NOT TRUE
-        ORDER BY tracking_date ASC LIMIT 1
-), (
-    SELECT new_value from gamification_karma_tracking new_tracking
-    WHERE new_tracking.user_id = gamification_karma_tracking.user_id
-        AND tracking_date::timestamp BETWEEN %(from_date)s AND %(to_date)s
-        AND consolidated IS NOT TRUE
-        ORDER BY tracking_date DESC LIMIT 1
-)
-FROM gamification_karma_tracking
-WHERE tracking_date::timestamp BETWEEN %(from_date)s AND %(to_date)s
-AND consolidated IS NOT TRUE
-GROUP BY user_id """
+        WITH old_tracking AS (
+            SELECT DISTINCT ON (user_id) user_id, old_value, tracking_date
+              FROM gamification_karma_tracking
+             WHERE tracking_date BETWEEN %(from_date)s
+               AND %(end_date)s
+               AND consolidated IS NOT TRUE
+          ORDER BY user_id, tracking_date ASC, id ASC
+        )
+            INSERT INTO gamification_karma_tracking (
+                            user_id,
+                            old_value,
+                            new_value,
+                            tracking_date,
+                            origin_ref,
+                            consolidated,
+                            reason)
+            SELECT DISTINCT ON (nt.user_id)
+                            nt.user_id,
+                            ot.old_value AS old_value,
+                            nt.new_value AS new_value,
+                            ot.tracking_date AS from_tracking_date,
+                            %(origin_ref)s AS origin_ref,
+                            TRUE,
+                            format(%(reason)s, ot.tracking_date::DATE, nt.tracking_date::DATE)
+              FROM gamification_karma_tracking AS nt
+              JOIN old_tracking AS ot
+                   ON ot.user_id = nt.user_id
+             WHERE nt.tracking_date BETWEEN %(from_date)s
+               AND %(end_date)s
+               AND nt.consolidated IS NOT TRUE
+          ORDER BY nt.user_id, nt.tracking_date DESC, id DESC
+        """
+
         self.env.cr.execute(select_query, {
             'from_date': from_date,
-            'to_date': end_date,
+            'end_date': end_date,
+            'origin_ref': f'res.users,{self.env.user.id}',
+            'reason': _('Consolidation from %s to %s'),
         })
-        results = self.env.cr.dictfetchall()
-        if results:
-            for result in results:
-                result['consolidated'] = True
-                result['tracking_date'] = fields.Date.to_string(from_date)
-            self.create(results)
-
-            self.search([
-                ('tracking_date', '>=', from_date),
-                ('tracking_date', '<=', end_date),
-                ('consolidated', '!=', True)]
-            ).unlink()
+
+        self.search([
+            ('tracking_date', '>=', from_date),
```

Comment:
Should we add an index on "tracking_date" ?

Seems like it's used a lot in the consolidate process.